File: libs/gomodels/channel_role.go
package model

import "github.com/lib/pq"

type ChannelRole struct {
	ID                        string          `gorm:"column:id;primary_key;type:uuid;default:uuid_generate_v4()" json:"id"`
	ChannelID                 string          `gorm:"column:channelId;type:uuid;" json:"-"`
	Name                      string          `gorm:"column:name;type:text;" json:"name"`
	Type                      ChannelRoleEnum `gorm:"column:type;type:text;" json:"type"`
	Permissions               pq.StringArray  `gorm:"column:permissions;type:text[]" json:"permissions"`
	RequiredWatchTime         int64           `gorm:"column:required_watch_time" json:"requiredWatchTime"`
	RequiredMessages          int32           `gorm:"column:required_messages" json:"requiredMessages"`
	RequiredUsedChannelPoints int64           `gorm:"column:required_used_channel_points" json:"requiredUsedChannelPoints"`

	Channel *Channels          `gorm:"foreignKey:ChannelID" json:"-"`
	Users   []*ChannelRoleUser `gorm:"foreignKey:RoleID" json:"-"`
}

func (ChannelRole) TableName() string {
	return "channels_roles"
}

type ChannelRoleEnum string

func (c ChannelRoleEnum) String() string {
	return string(c)
}

const (
	ChannelRoleTypeBroadcaster ChannelRoleEnum = "BROADCASTER"
	ChannelRoleTypeModerator   ChannelRoleEnum = "MODERATOR"
	ChannelRoleTypeSubscriber  ChannelRoleEnum = "SUBSCRIBER"
	ChannelRoleTypeVip         ChannelRoleEnum = "VIP"
	ChannelRoleTypeCustom      ChannelRoleEnum = "CUSTOM"
)

type RolePermissionEnum string

func (r RolePermissionEnum) String() string {
	return string(r)
}

const (
	RolePermissionCanAccessDashboard RolePermissionEnum = "CAN_ACCESS_DASHBOARD"

	RolePermissionUpdateChannelTitle    RolePermissionEnum = "UPDATE_CHANNEL_TITLE"
	RolePermissionUpdateChannelCategory RolePermissionEnum = "UPDATE_CHANNEL_CATEGORY"

	RolePermissionViewCommands   RolePermissionEnum = "VIEW_COMMANDS"
	RolePermissionManageCommands RolePermissionEnum = "MANAGE_COMMANDS"

	RolePermissionViewKeywords   RolePermissionEnum = "VIEW_KEYWORDS"
	RolePermissionManageKeywords RolePermissionEnum = "MANAGE_KEYWORDS"

	RolePermissionViewTimers   RolePermissionEnum = "VIEW_TIMERS"
	RolePermissionManageTimers RolePermissionEnum = "MANAGE_TIMERS"

	RolePermissionViewIntegrations   RolePermissionEnum = "VIEW_INTEGRATIONS"
	RolePermissionManageIntegrations RolePermissionEnum = "MANAGE_INTEGRATIONS"

	RolePermissionViewSongRequests   RolePermissionEnum = "VIEW_SONG_REQUESTS"
	RolePermissionManageSongRequests RolePermissionEnum = "MANAGE_SONG_REQUESTS"

	RolePermissionViewModeration   RolePermissionEnum = "VIEW_MODERATION"
	RolePermissionManageModeration RolePermissionEnum = "MANAGE_MODERATION"

	RolePermissionViewVariables   RolePermissionEnum = "VIEW_VARIABLES"
	RolePermissionManageVariables RolePermissionEnum = "MANAGE_VARIABLES"

	RolePermissionViewGreetings   RolePermissionEnum = "VIEW_GREETINGS"
	RolePermissionManageGreetings RolePermissionEnum = "MANAGE_GREETINGS"

	RolePermissionViewOverlays   RolePermissionEnum = "VIEW_OVERLAYS"
	RolePermissionManageOverlays RolePermissionEnum = "MANAGE_OVERLAYS"
)

File: libs/typeorm/src/entities/ChannelRole.ts
/* eslint-disable import/no-cycle */
import { Column, Entity, JoinColumn, ManyToOne, OneToMany, PrimaryGeneratedColumn } from 'typeorm';

import { Channel } from './Channel';
import { ChannelRoleUser } from './ChannelRoleUser';

export enum RoleType {
  BROADCASTER = 'BROADCASTER',
  MODERATOR = 'MODERATOR',
  SUBSCRIBER = 'SUBSCRIBER',
  VIP = 'VIP',
  CUSTOM = 'CUSTOM',
}

export enum RoleFlags {
  CAN_ACCESS_DASHBOARD = 'CAN_ACCESS_DASHBOARD',

  UPDATE_CHANNEL_TITLE = 'UPDATE_CHANNEL_TITLE',
  UPDATE_CHANNEL_CATEGORY = 'UPDATE_CHANNEL_CATEGORY',

  VIEW_COMMANDS = 'VIEW_COMMANDS',
  MANAGE_COMMANDS = 'MANAGE_COMMANDS',

  VIEW_KEYWORDS = 'VIEW_KEYWORDS',
  MANAGE_KEYWORDS = 'MANAGE_KEYWORDS',

  VIEW_TIMERS = 'VIEW_TIMERS',
  MANAGE_TIMERS = 'MANAGE_TIMERS',

  VIEW_INTEGRATIONS = 'VIEW_INTEGRATIONS',
  MANAGE_INTEGRATIONS = 'MANAGE_INTEGRATIONS',

  VIEW_SONG_REQUESTS = 'VIEW_SONG_REQUESTS',
  MANAGE_SONG_REQUESTS = 'MANAGE_SONG_REQUESTS',

  VIEW_MODERATION = 'VIEW_MODERATION',
  MANAGE_MODERATION = 'MANAGE_MODERATION',

  VIEW_VARIABLES = 'VIEW_VARIABLES',
  MANAGE_VARIABLES = 'MANAGE_VARIABLES',

  VIEW_GREETINGS = 'VIEW_GREETINGS',
  MANAGE_GREETINGS = 'MANAGE_GREETINGS',
}

@Entity('channels_roles')
export class ChannelRole {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @ManyToOne(() => Channel, channel => channel.roles)
  @JoinColumn({ name: 'channelId' })
  channel?: Channel;

  @Column()
  channelId: string;

  @Column()
  name: string;

  @Column('enum', { enum: RoleType, default: RoleType.CUSTOM })
  type: RoleType;

  @Column({
    type: 'enum',
    enum: RoleFlags,
    array: true,
    default: [],
  })
  permissions: RoleFlags[];

  @OneToMany(() => ChannelRoleUser, _ => _.role)
  users?: ChannelRoleUser[];
}